// crochess-api/dist/moveNotation.js
const moveNotation = (notation) => {
    function castle(side) {
        notation = side === 'kingside' ? '0-0' : '0-0-0';
        return notation;
    }
    const affix = {
        pieceNotation: (pieceType, differentiation) => {
            differentiation = differentiation || '';
            switch (pieceType) {
                case 'pawn': {
                    notation = differentiation + notation;
                    return notation;
                }
                case 'king': {
                    notation = 'K' + notation;
                    return notation;
                }
                case 'knight': {
                    notation = 'N' + differentiation + notation;
                    return notation;
                }
                default: {
                    notation = pieceType[0].toUpperCase() + differentiation + notation;
                    return notation;
                }
            }
        },
        capture: () => {
            notation = `x${notation}`;
            return notation;
        },
        promote: (pieceType) => {
            const suffix = '=' + pieceType === 'knight' ? 'N' : pieceType.charAt(0).toUpperCase();
            notation += '=' + suffix;
            return notation;
        },
        check: () => {
            notation += '+';
            return notation;
        },
        checkmate: () => {
            notation += '#';
            return notation;
        }
    };
    function get(moveDetails) {
        const { capture, castle: castleSide, promote, check, checkmate, pieceType, differentiation } = moveDetails;
        if (castleSide)
            return castle(castleSide);
        if (capture)
            affix.capture();
        affix.pieceNotation(pieceType, differentiation);
        if (promote)
            affix.promote(promote);
        if (checkmate)
            affix.checkmate();
        else if (check)
            affix.check();
        return notation;
    }
    return {
        get
    };
};
export default moveNotation;
